wasm/src/lsp/types.rs

```diff
@@ -352,3 +355,138 @@ fn test_doc_range_overlap_at_zero() {
+                if position.character > 0 && (position.character as usize) <= line.len() {
+                    let prev_char = position.character - 1;
+                    let the_char = line[prev_char as usize];
+                    Some((
+                        the_char,
+                        Position {
+                            line: position.line,
+                            character: prev_char,
+                        },
+                    ))
+                } else {
+                    None
+                }
+            })
+        }
+    }
+}
+
+// Note: This is using a directive that ensures that this code is only included
+// in the test build.  It is not necessary to be concerned that it will be
+// included in another configuration.  Although it's name contains test, it will
+// not be used as a test in this language.
+#[cfg(test)]
+fn make_test_doc_data_object_for_the_subsequent_test_code_1() -> DocData {
+    let mut comment_hashmap = HashMap::new();
+    // There is a comment at line 2, column 24 and at line 3, column 8
+    comment_hashmap.insert(1, 23);
+    comment_hashmap.insert(2, 7);
+    // vr vec Rc
+    let vr = |s: &str| {
+        let bv: Vec<u8> = s.as_bytes().to_vec();
+        Rc::new(bv)
+    };
+    DocData {
+        // Note: this structure contains owned objects such as strings and vecs.
+        // The ownership is intended.
+        fullname: "test_name.clsp".to_string(),
+        // Note: This vec contains reference counted lines so patches can move
+        // them without cloning them.  The Rc here, as well as the string
+        // ownership are intended.
+        text: vec![
+            vr("(mod (X)"),
+            vr(" (defun F (A) (+ A 1)) ;; A function"),
+            vr(" (F X) ;; Call"),
+            vr(" )")
+        ],
+        version: 1,
+        comments: comment_hashmap
+    }
+}
+
+// Note: This is a test in this language.  Readers find non-segregated tests an
+// easier read so tests are inline.
+#[test]
+fn test_doc_data_nth_line_ref_1() {
+    let dd = make_test_doc_data_object_for_the_subsequent_test_code_1();
+    assert_eq!(dd.nth_line_ref(2), Some(&" (F X) ;; Call".as_bytes().to_vec()));
+}
+
+#[test]
+fn test_doc_data_nth_line_ref_2() {
+    let dd = make_test_doc_data_object_for_the_subsequent_test_code_1();
+    assert_eq!(dd.nth_line_ref(5), None);
+}
+
+#[test]
+fn test_doc_data_get_prev_position_1() {
+    let dd = make_test_doc_data_object_for_the_subsequent_test_code_1();
+    // Zero based.
+    let mut position = Some(Position { line: 3, character: 2 });
+    let mut all_expected_characters = Vec::new();
+    let mut got_characters = Vec::new();
+    let mut have_line_jumps = Vec::new();
+    let want_line_jumps = vec![2, 16, 52];
+
+    for l in dd.text.iter().rev() {
+        let borrowed_line: &Vec<u8> = l.borrow();
+        let mut reversed_line = borrowed_line.iter().rev().copied().collect();
```

Comment:
As I suggested above, `borrow()` here is not required.
```rust
        let borrowed_line: &Vec<u8> = l.borrow();
        let mut reversed_line = borrowed_line.iter().rev().copied().collect();
```
can be simplified to
```rust
let mut reversed_line = l.iter().rev().copied().collect();
```

Reply:
because many things use slice references, and we actually want a vec ref here, the types can be ambiguous.  I do this because it's uniform.